remote_attestation/rust/src/attestation/unattested_session.rs

```diff
@@ -0,0 +1,146 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::proto::{AttestationIdentity, AttestationInfo, AttestationReport};
+use anyhow::{anyhow, Context};
+
+/// Represents an attestation process, where current machine remotely attests to a remote peer and
+/// sends attestation info to it.
+pub struct UnattestedSelf {
+    /// PEM encoded X.509 certificate that signs TEE firmware key.
+    tee_certificate: Vec<u8>,
+}
+
+impl UnattestedSelf {
+    pub fn new(tee_certificate: &[u8]) -> Self {
+        Self {
+            tee_certificate: tee_certificate.to_vec(),
+        }
+    }
+}
+
+/// Represents an attestation process, where current machine remotely attests a remote peer and
+/// verifies its attestation info.
+pub struct UnattestedPeer {
+    /// Expected value of the peer's TEE measurement.
+    expected_tee_measurement: Vec<u8>,
+}
+
+impl UnattestedPeer {
+    pub fn new(expected_tee_measurement: &[u8]) -> Self {
+        Self {
+            expected_tee_measurement: expected_tee_measurement.to_vec(),
+        }
+    }
+}
+
+/// Represents an attestation process, where current machine and a remote peer remotely attest each
+/// other.
+pub struct MutuallyUnattested {
+    /// PEM encoded X.509 certificate that signs TEE firmware key.
+    tee_certificate: Vec<u8>,
+    /// Expected value of the peer's TEE measurement.
+    expected_tee_measurement: Vec<u8>,
+}
+
+impl MutuallyUnattested {
+    pub fn new(tee_certificate: &[u8], expected_tee_measurement: &[u8]) -> Self {
+        Self {
+            tee_certificate: tee_certificate.to_vec(),
+            expected_tee_measurement: expected_tee_measurement.to_vec(),
+        }
+    }
+}
+
+pub trait UnattestedSession {
+    fn get_attestation_info(&self, public_key_hash: &[u8]) -> Option<AttestationInfo>;
+    fn verify_attestation(self, peer_identity: &AttestationIdentity) -> anyhow::Result<()>;
+}
+
+impl UnattestedSession for UnattestedSelf {
+    fn get_attestation_info(&self, public_key_hash: &[u8]) -> Option<AttestationInfo> {
+        // Generate attestation info with a TEE report.
+        // TEE report contains a hash of the public key.
+        Some(AttestationInfo {
+            report: Some(AttestationReport::new(&public_key_hash)),
+            certificate: self.tee_certificate.clone(),
+        })
+    }
+
+    fn verify_attestation(self, _peer_identity: &AttestationIdentity) -> anyhow::Result<()> {
+        Ok(())
```

Comment:
I'm not sure whether this is the final version or just a placeholder, or anyways what this method is supposed to return. Could you add more comments on the trait definition to explain what the methods do and how they are supposed to be implemented and by whom?